lib/Sanger/CGP/NgsQc/ValidateSampleMetaAndBamHeader.pm

```diff
@@ -0,0 +1,755 @@
+          if (defined $cl) {
+            my @words = split(/\s+/, $cl);
+            my %valid_words;
+            my $count = 0;
+            for (my $i = 0; $i < scalar @words; $i++) {
+              if ($words[$i] eq '-K' && ! exists $valid_words{'-K'}) {
+                $valid_words{$words[$i]} = 1;
+                if ($words[$i + 1] == $valid_pg_hash{CL}{'-K'}) {
+                  $count += 1;
+                }
+              } else {
+                if (! exists $valid_words{$words[$i]} && exists $valid_pg_hash{CL}{$words[$i]}) {
+                  $valid_words{$words[$i]} = 1;
+                  $count += 1;
+                }
+              }
+            }
+            if ($count == 6) {
+              $match_pp_remapped_sq_pg = 1;
+            } else {
+              warn "bwa command line in head does not match pp_remapped_bam pg line!\n";
+              $match_pp_remapped_sq_pg = 0; last;
+            }
+          } else {
+            warn "PG CL is not found!\n";
+            $match_pp_remapped_sq_pg = 0; last;
+          }
+        }
+      } else {
+        warn "No valid PG BWA header lines!\n";
+        $match_pp_remapped_sq_pg = 0;
+      }
+    } else {
+      warn "No valid PG header lines / invalid SQ lines\n";
+      $match_pp_remapped_sq_pg = 0;
+    }
+
+    if ($match_pp_remapped_sq_pg == 1) {
+      warn "this is a pp_remapped_bam!\n";
+    } else {
+      warn "this is a raw_bam!\n";
+    }
+
+    my $bam = $sam->hts_file;
+    my $header = $bam->header_read;
+    my $processed_x = 0;
+    my $start = time;
+    my $processed_x_mill = 0;
+
+    my %reported_error;
+
+    while (my $a = $bam->read1($header)) {
+      $processed_x += 1;
+
+      my ($rg) = ($a->get_tag_values('RG'));
+      if (defined $rg) {
+        if (exists $headers{'RG'}{$rg}) {
+          if ($headers{'RG'}{$rg} == 0) {
+            $headers{'RG'}{$rg} = 1;
+          }
+        } else {
+          if (! exists $reported_error{$FAIL_CODE_120}) {
+            warn "found a read with a RG tag not in header.\n";
+            $reported_error{$FAIL_CODE_120} = 1;
+            $bam_state .= $FAIL_CODE_120."; ";
+          }
+        }
+      } else {
+        if (! exists $reported_error{$FAIL_CODE_110}) {
+          warn "found a read with no RG tag.\n";
+          $reported_error{$FAIL_CODE_110} = 1;
+          $bam_state .= $FAIL_CODE_110."; ";
+        }
+      }
+
+      if($processed_x % $count_base_number == 0) {
+        $processed_x_mill++;
+        my $end = time;
+        my $elapsed = $end - $start;
+        $start = $end;
+        warn "$processed_x_mill million ($processed_x) reads processed [time elapsed: ${elapsed}s].\n";
+
+        my $not_found = '';
+        for my $rg_header (keys %{$headers{'RG'}}) {
+          if ($headers{'RG'}{$rg_header} == 0) {
+            $not_found .= ' '.$rg_header;
+          }
+        }
+
+        $not_found =~ s/^ //;
+        if ($not_found ne '') {
```

Comment:
in array you can test `scalar @not_found` instead, string operations are expensive.